JS/Lesson18/tugba/recursion.js

```diff
@@ -13,130 +13,316 @@
-		city: 'New York',
-		country: 'USA',
+	street: '123 Main St',
+	city: 'New York',
+	country: 'USA',
 	},
 	hobbies: ['reading', 'coding'],
 	social: {
-		twitter: '@johndoe',
-		linkedIn: 'johndoe',
+	twitter: '@johndoe',
+	linkedIn: 'johndoe',
 	},
-};
-// 4. Write a recursive function to count the number of vowels in a string.
-const sentenceEx4 = 'Hello, how are you?';
-const vowels = ['a', 'e', 'i', 'o', 'u'];
-
-// 5. Write a recursive function to remove all occurrences of a specified character from a string.
-const sentenceEx5 = 'Hello, how are you?';
-const charToRemove = 'o';
-
-// 6. Write a recursive function to check if an array includes a specific value.
-const numbers = [1, 2, 3, 4, 5];
-const valueToCheck = 3;
-
-// 7. Write a recursive function to find the maximum depth of a nested object.
-const nestedObjEx7 = {
+	};
+	
+	function flattenObject(obj){
+	let flattened = {};
+	
+	function flatten(obj, prefix = " "){
+	for(let key in obj){
+	if(obj.hasOwnProperty(key)){
+	let propName = "";
+	if(prefix){
+	propName = prefix + "." + key;
+	}
+	else{
+	propName = prefix;
+	}
+	if(typeof obj[key] === "object" && obj[key] !== null){
+	flatten(obj[key], propName);
+	}
+	else{
+	flattened[propName] = obj[key];
+	}
+	}
+	}
+	}
+	flatten(obj);
+	return flattened;
+	}
+	
+	console.log(flattenObject(nestedObj));
+	
+	console.log("================================")
+	// 4. Write a recursive function to count the number of vowels in a string.
+	console.log("Example4: ");
+	
+	const sentenceEx4 = 'Hello, how are you?';
+	const vowels = ['a', 'e', 'i', 'o', 'u'];
+	
+	function countVowelsRecursive(str, vowels) {
+	if (str.length === 0) {
+	 return 0;
+	}
+	
+	const firstChar = str[0].toLowerCase();
+	const isVowel = vowels.includes(firstChar);
+	 
+	let count = 0;
+	if (isVowel) {
+	 count = 1;
+	} else {
+	 count = 0;
+	}
+	count += countVowelsRecursive(str.slice(1), vowels);
+	
+	return count;
+	}
+	
+	const vowelCount = countVowelsRecursive(sentenceEx4, vowels);
+	console.log('Number of vowels:', vowelCount);
+	
+	console.log("================================")
+	// 5. Write a recursive function to remove all occurrences of a specified character from a string.
+	console.log("Example5: ");
+	
+	const sentenceEx5 = 'Hello, how are you?';
+	const charToRemove = 'o';
+	
+	function removeSpecified(sentence, charToRemove){
+	if(sentence.length === 0){
+	return "";
+	}
+	
+	const firstChar = sentence[0];
+	
+	if(firstChar === charToRemove){
+	return removeSpecified(sentence.slice(1), charToRemove);
+	}
+	else{
+	return firstChar + removeSpecified(sentence.slice(1), charToRemove);
+	}
+	}
+	
+	const result = removeSpecified(sentenceEx4, charToRemove);
+	console.log(result);
+	
+	console.log("================================")
+	// 6. Write a recursive function to check if an array includes a specific value.
+	console.log("Example6: ");
+	const numbers = [1, 2, 3, 4, 5];
+	const valueToCheck = 3;
+	
+	function removeSpecified(sentence, charToRemove){
```

Comment:
A function with name 'removeSpecified' is already declared in the same scope (ex. 5) so you cannot use it here.
Instead, the correct solution to Ex. 6 would be:
```
function includesValue(array, value) {
    // Base case: if the array is empty, the value is not found
    if (array.length === 0) {
        return false;
    }

    // Check if the first element is the value we're looking for
    if (array[0] === value) {
        return true;
    }

    // Recursive call on the rest of the array
    return includesValue(array.slice(1), value);
}

console.log(includesValue(numbers, valueToCheck));
```